[GStreamer][MediaStream] Racy deadlock upon track removal request

philn opened the pull request:
#### 262bd3311741fd533e81ff88bec3120e64ed232f
<pre>
[GStreamer][MediaStream] Racy deadlock upon track removal request
<a href="https://bugs.webkit.org/show_bug.cgi?id=244288">https://bugs.webkit.org/show_bug.cgi?id=244288</a>

Reviewed by Xabier Rodriguez-Calvar.

EOS is now sent from the appsrc streaming thread, in case it was requested following a MediaStream
endTrack or track removal notification from WebCore. Using this approach the main thread and
streaming thread can no longer deadlock.

This patch also includes several drive-by changes, notably resetting the flow combiner when going
from PAUSED to READY and also when a RECONFIGURE event is received. This element now also fully
takes advantage of the GstObjectLocker RAII facilities.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
(WebKitMediaStreamObserver::didRemoveTrack):
(webkitMediaStreamSrcDispose):
(webkitMediaStreamSrcChangeState):
(webkitMediaStreamSrcChain):
(webkitMediaStreamSrcPostStreamCollection):
(webkitMediaStreamSrcAddPad):
(webkitMediaStreamSrcTrackEnded): Deleted.
(InternalSource::trackEnded): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::~RealtimeOutgoingMediaSourceGStreamer):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stop):

Canonical link: <a href="https://commits.webkit.org/253940@main">https://commits.webkit.org/253940@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f25fce2cad816e691c64299a5d020188579b9631

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87552 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31642 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18334 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96779 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/150538 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91525 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30005 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/26158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79671 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91533 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93171 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/24239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74327 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24071 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79203 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79413 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67088 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27712 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13246 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27666 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14262 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2773 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29356 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37135 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29277 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33540 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
```diff
@@ -597,9 +645,42 @@ static void webkit_media_stream_src_class_init(WebKitMediaStreamSrcClass* klass)
+        gst_element_set_state(appSrc, GST_STATE_NULL);
+        gst_bin_remove(GST_BIN_CAST(self), appSrc);
+        gst_element_set_locked_state(appSrc, false);
+        GST_STATE_UNLOCK(element.get());
```
Why don't you use a new scope and the automatic locker?

philn replied:
Because lazy

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.cpp:
```diff
@@ -68,9 +68,11 @@ RealtimeOutgoingMediaSourceGStreamer::~RealtimeOutgoingMediaSourceGStreamer()
     gst_element_set_state(m_bin.get(), GST_STATE_NULL);
     if (auto pipeline = adoptGRef(gst_element_get_parent(m_bin.get())))
         gst_bin_remove(GST_BIN_CAST(pipeline.get()), m_bin.get());
+    gst_element_set_locked_state(m_bin.get(), FALSE);
```
It might be interesting to write an automatic locker for this one as well. No strong opinion.

philn replied:
Maybe later!

philn committed:
[GStreamer][MediaStream] Racy deadlock upon track removal request
https://bugs.webkit.org/show_bug.cgi?id=244288

Reviewed by Xabier Rodriguez-Calvar.

EOS is now sent from the appsrc streaming thread, in case it was requested following a MediaStream
endTrack or track removal notification from WebCore. Using this approach the main thread and
streaming thread can no longer deadlock.

This patch also includes several drive-by changes, notably resetting the flow combiner when going
from PAUSED to READY and also when a RECONFIGURE event is received. This element now also fully
takes advantage of the GstObjectLocker RAII facilities.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
(WebKitMediaStreamObserver::didRemoveTrack):
(webkitMediaStreamSrcDispose):
(webkitMediaStreamSrcChangeState):
(webkitMediaStreamSrcChain):
(webkitMediaStreamSrcPostStreamCollection):
(webkitMediaStreamSrcAddPad):
(webkitMediaStreamSrcTrackEnded): Deleted.
(InternalSource::trackEnded): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::~RealtimeOutgoingMediaSourceGStreamer):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::stop):

Canonical link: https://commits.webkit.org/253940@main
